autosklearn/pipeline/base.py

```diff
@@ -19,13 +19,15 @@ class BasePipeline(Pipeline):
 
     def __init__(self, config=None, steps=None, dataset_properties=None,
                  include=None, exclude=None, random_state=None,
-                 init_params=None):
+                 init_params=None, target_type=None):
```

Comment:
Hello, thanks a lot for the enhancements, they look pretty good.

This is the only change I am not fully able to understand. Do you think we can instead of defining another target_type, get it on the dataset properties?

For example, we define that [here](https://github.com/automl/auto-sklearn/blob/275d0d6b20d16822252d8b50bf71b1c787187f09/autosklearn/pipeline/classification.py#L161). I am wondering in which scenarios is the `target_type` missing?

Reply:
In recent changes, to validate include/exclude params exhaustively, we are also validating for the components for the steps in these params. To validate against all [available components](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/base.py#L486), `target_type` is required in `dataset_properties` for [feature preprocessors](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/components/feature_preprocessing/__init__.py#L62)  and [data preprocessors](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/pipeline/components/data_preprocessing/__init__.py#L63) 

Now `target_type` is required in `dataset_properties` while creating pipeline objects and it is missing at all places ([utils/pipeline.py](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/util/pipeline.py#L82) as well as [abstract_evaluator.py](https://github.com/rabsr/auto-sklearn/blob/config_data_preprocess/autosklearn/evaluation/abstract_evaluator.py#L284)). Instead of adding it at various places and then validating and updating when creating pipeline object, I tried to have minimum changes to achieve the same.

Please do let me know if you have better approach to handle this. I will improve on implementation.

Reply:
Thanks a lot for the explanation.

Yeah, this is clearly needed when you instantiate the object, and in particular when dataset_properties is `None`/missing this target_type.

I do have one implementation change request to achieve the same goal. Can you add this change [here](https://github.com/automl/auto-sklearn/blob/81115f6d1184966c872a82403a3c6abc60cef9db/autosklearn/pipeline/classification.py#L74) rather than a new attribute?

Something like this:
```
if dataset_properties is None or 'target_type' not in dataset_properties:
    --> then add it here in the classification/regression pipeline
```

Why not an argument to the pipeline? This complicates things with scikit learn, for example with `get_params`/`set_params` you need all attributes to be stored as `self.ATTRIBUTE_NAME`. Also, we are trying to keep the interface the same if possible.

Reply:
Thanks for clarifying. Have updated as per your suggestion.